File: src/core/validators/Subdomain/SubdomainParentDomainNameMustMatchADomain.cs
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Subdomain
{
    public class SubdomainParentDomainNameMustMatchADomain : ValidationRuleBase<MetaEdGrammar.SubdomainContext>
    {
        private readonly ISymbolTable _symbolTable;

        public SubdomainParentDomainNameMustMatchADomain(ISymbolTable symbolTable)
        {
            _symbolTable = symbolTable;
        }

        public override bool IsValid(MetaEdGrammar.SubdomainContext context)
        {
            var parentDomainName = context.parentDomainName().IdText();

            var domainType = MetaEdGrammar.TokenName(MetaEdGrammar.DOMAIN);

            return _symbolTable.IdentifierExists(domainType, parentDomainName);
        }

        public override string GetFailureMessage(MetaEdGrammar.SubdomainContext context)
        {
            return string.Format("Subdomain '{0}' is part of '{1}' which does not match any declared domain.", context.EntityName(), context.parentDomainName().IdText());
        }
    }
}

File: src/core/validators/Subdomain/SubdomainParentDomainNameMustMatchADomain.js
// @flow
import type SymbolTable from '../SymbolTable';
import { errorRuleBase } from '../ValidationRuleBase';
import { includeRuleBase } from '../ValidationRuleRepository';
import { MetaEdGrammar } from '../../../grammar/gen/MetaEdGrammar';
import SymbolTableEntityType from '../SymbolTableEntityType';

function valid(ruleContext: any, symbolTable: SymbolTable): boolean {
  return symbolTable.identifierExists(SymbolTableEntityType.domain(), ruleContext.parentDomainName().ID().getText());
}

// eslint-disable-next-line no-unused-vars
function failureMessage(ruleContext: any, symbolTable: SymbolTable): string {
  return `Subdomain '${ruleContext.subdomainName().ID().getText()}' is part of '${ruleContext.parentDomainName().ID().getText()}' which does not match any declared domain.`;
}

const validationRule = errorRuleBase(valid, failureMessage);
// eslint-disable-next-line import/prefer-default-export
export const includeRule = includeRuleBase(MetaEdGrammar.RULE_subdomain, validationRule);
